packages/fxa-auth-server/lib/routes/emails.js

```diff
@@ -586,108 +558,92 @@ module.exports = (
-          return db
-            .getSecondaryEmail(email)
-            .then(secondaryEmailRecord => {
-              if (secondaryEmailRecord.isPrimary) {
-                if (secondaryEmailRecord.isVerified) {
-                  throw error.verifiedPrimaryEmailAlreadyExists();
-                }
-
-                const msSinceCreated =
-                  Date.now() - secondaryEmailRecord.createdAt;
-                const minUnverifiedAccountTime =
-                  config.secondaryEmail.minUnverifiedAccountTime;
-                if (msSinceCreated >= minUnverifiedAccountTime) {
-                  return db.deleteAccount(secondaryEmailRecord).then(() =>
-                    log.info('accountDeleted.unverifiedSecondaryEmail', {
-                      ...secondaryEmailRecord,
-                    })
-                  );
-                } else {
-                  throw error.unverifiedPrimaryEmailNewlyCreated();
-                }
-              }
+        if (sessionToken.tokenVerificationId) {
+          throw error.unverifiedSession();
+        }
 
-              // Only delete secondary email if it is unverified and does not belong
-              // to the current user.
-              if (
-                !secondaryEmailRecord.isVerified &&
-                !butil.buffersAreEqual(secondaryEmailRecord.uid, uid)
-              ) {
-                return db.deleteEmail(
-                  secondaryEmailRecord.uid,
-                  secondaryEmailRecord.email
-                );
-              }
-            })
-            .catch(err => {
-              if (err.errno !== error.ERRNO.SECONDARY_EMAIL_UNKNOWN) {
-                throw err;
-              }
-            });
+        if (sessionToken.email.toLowerCase() === email.toLowerCase()) {
+          throw error.yourPrimaryEmailExists();
         }
 
-        function generateRandomValues() {
-          return random.hex(16).then(hex => {
-            emailData.emailCode = hex;
+        await deleteAccountIfUnverified();
+
+        const hex = await random.hex(16);
+        emailData.emailCode = hex;
+
+        await db.createEmail(uid, emailData);
+
+        const geoData = request.app.geo;
+        try {
+          await mailer.sendVerifySecondaryEmail([emailData], sessionToken, {
+            code: emailData.emailCode,
+            deviceId: sessionToken.deviceId,
+            acceptLanguage: request.app.acceptLanguage,
+            email: emailData.email,
+            primaryEmail,
+            ip,
+            location: geoData.location,
+            timeZone: geoData.timeZone,
+            uaBrowser: sessionToken.uaBrowser,
+            uaBrowserVersion: sessionToken.uaBrowserVersion,
+            uaOS: sessionToken.uaOS,
+            uaOSVersion: sessionToken.uaOSVersion,
+            uid,
           });
+        } catch (err) {
+          log.error('mailer.sendVerifySecondaryEmail', { err: err });
+          await db.deleteEmail(emailData.uid, emailData.normalizedEmail);
```

Comment:
What happens if db.deleteEmail fails here? 500 error?

Reply:
That appears to be the case.  Since the next line is throwing an error, it's also a 500 if `db.deleteEmail` succeeds.